src/typechecker.ts

```diff
@@ -0,0 +1,896 @@
+    [a, b] = [b, a];
+  }
+  if (a.kind === 'union') {
+    if (b.kind === 'union') {
+      return b.of.reduce(
+        (acc: Type, t) => (acc.kind === 'union' ? addToUnion(acc.of, t) : join(acc, t)),
+        a
+      );
+    }
+    return addToUnion(a.of, b);
+  }
+  if (
+    (a.kind === 'list' && b.kind === 'list') ||
+    (a.kind === 'completion' && b.kind === 'completion')
+  ) {
+    return { kind: a.kind, of: join(a.of, b.of) };
+  }
+  return { kind: 'union', of: [a, b as NonUnion] };
+}
+
+export function meet(a: Type, b: Type): Type {
+  if (dominates(a, b)) {
+    return b;
+  }
+  if (dominates(b, a)) {
+    return a;
+  }
+  if (a.kind !== 'union' && b.kind === 'union') {
+    [a, b] = [b, a];
+  }
+  if (a.kind === 'union') {
+    // union is join. meet distributes over join.
+    return a.of.map(t => meet(t, b)).reduce(join);
+  }
+  if (
+    (a.kind === 'list' && b.kind === 'list') ||
+    (a.kind === 'completion' && b.kind === 'completion')
+  ) {
+    return { kind: a.kind, of: meet(a.of, b.of) };
+  }
+  return { kind: 'never' };
+}
+
+function serialize(type: Type): string {
+  switch (type.kind) {
+    case 'unknown': {
+      return 'unknown';
+    }
+    case 'never': {
+      return 'never';
+    }
+    case 'union': {
+      const parts = type.of.map(serialize);
+      if (parts.length > 2) {
+        return parts.slice(0, -1).join(', ') + ', or ' + parts[parts.length - 1];
+      }
+      return parts[0] + ' or ' + parts[1];
+    }
+    case 'list': {
+      if (type.of.kind === 'never') {
+        return 'empty List';
+      }
+      return 'List of ' + serialize(type.of);
+    }
+    case 'record': {
+      return 'Record';
+    }
+    case 'completion': {
+      if (type.of.kind === 'never') {
+        return 'an abrupt Completion Record';
+      } else if (type.of.kind === 'unknown') {
+        return 'a Completion Record';
+      }
+      return 'a Completion Record normally holding ' + serialize(type.of);
+    }
+    case 'real': {
+      return 'mathematical value';
+    }
+    case 'integer':
+    case 'non-negative integer':
+    case 'negative integer':
+    case 'positive integer':
+    case 'null':
+    case 'undefined': {
+      return type.kind;
+    }
+    case 'concrete string': {
+      return `"${type.value}"`;
+    }
+    case 'concrete real': {
+      return type.value;
+    }
+    case 'concrete boolean': {
+      return `${type.value}`;
+    }
+    case 'enum value': {
+      return `~${type.value}~`;
+    }
+    case 'concrete number': {
+      if (Object.is(type.value, 0 / 0)) {
+        return '*NaN*';
+      }
+      let repr;
+      if (Object.is(type.value, -0)) {
+        repr = '-0';
+      } else if (type.value === 0) {
+        repr = '+0';
+      } else if (type.value === 2e308) {
+        repr = '+&infin;';
+      } else if (type.value === -2e308) {
+        repr = '-&infin;';
+      } else if (type.value > 4503599627370495.5) {
+        repr = String(BigInt(type.value));
```

Comment:
This will throw if you pass an excessively large float. The parser shouldn't throw. It can surface an error to the user, but it shouldn't throw.

Reply:
Will it? What's an example?

```
> String(BigInt(1e308))
< '100000000000000001097906362944045541740492309677311846336810682903157585404911491537163328978494688899061249669721172515611590283743140088328307009198146046031271664502933027185697489699588559043338384466165001178426897626212945177628091195786707458122783970171784415105291802893207873272974885715430223118336'
```

Reply:
This is the largest non-integral double.

Reply:
Oh, fair enough. WFM then.